test/unit/assorted/server_discovery_and_monitoring.spec.test.ts

```diff
@@ -157,129 +167,109 @@ function cloneMap(map) {
 
-        const incompatibilityExpected = phase.outcome ? !phase.outcome.compatible : false;
-        if (incompatibilityExpected) {
-          topology.on('error', incompatibilityHandler);
+    if (phase.responses) {
+      for (const [address, hello] of phase.responses) {
+        topology.serverUpdateHandler(new ServerDescription(address, hello));
+      }
+      if (phase.outcome) {
+        assertOutcomeExpectations(topology, events, phase.outcome);
+        if (phase.outcome.compatible === false) {
+          for (const errorEvent of errorEvents) {
+            expect(errorEvent).to.be.instanceOf(MongoCompatibilityError);
+            expect(errorEvent.message).to.match(/but this version of the driver/);
+          }
         }
+      }
 
-        // if (phase.description) {
-        //   console.log(`[phase] ${phase.description}`);
-        // }
-
-        if (phase.responses) {
-          // simulate each hello response
-          phase.responses.forEach(response =>
-            topology.serverUpdateHandler(new ServerDescription(response[0], response[1]))
-          );
-          phaseDone();
-        } else if (phase.applicationErrors) {
-          eachAsyncSeries(
-            phase.applicationErrors,
-            (appError, phaseCb) => {
-              let withConnectionStub = sinon
-                .stub(ConnectionPool.prototype, 'withConnection')
-                .callsFake(withConnectionStubImpl(appError));
-
-              const server = topology.s.servers.get(appError.address);
-              server.command(ns('admin.$cmd'), { ping: 1 }, undefined, err => {
-                expect(err).to.exist;
-                withConnectionStub.restore();
-
-                phaseCb();
-              });
-            },
-            err => {
-              expect(err).to.not.exist;
-              phaseDone();
-            }
+      topology.removeAllListeners('error');
+      events = [];
+    } else if (phase.applicationErrors) {
+      for (const appError of phase.applicationErrors) {
+        const withConnectionStub = sinon
+          .stub(ConnectionPool.prototype, 'withConnection')
+          .callsFake(withConnectionStubImpl(appError));
+
+        const server = topology.s.servers.get(appError.address);
+        const res = promisify(server.command.bind(server))(ns('admin.$cmd'), { ping: 1 }, {});
+        withConnectionStub.restore();
+
+        const thrownError = await res.catch(error => error);
+        const isApplicationError = error => {
+          // These errors all come from the withConnection stub
+          return (
+            error instanceof MongoNetworkError ||
+            error instanceof MongoNetworkTimeoutError ||
+            error instanceof MongoServerError
           );
-        } else {
-          phaseDone();
-        }
-      },
-      err => {
-        expect(err).to.not.exist;
-        done();
+        };
+        expect(thrownError).to.satisfy(isApplicationError);
```

Comment:
can we add a custom message here for the expect fail case?

Reply:
Done!